File: nodecore-miners-pop/src/main/java/nodecore/miners/pop/rules/actions/MineAction.java
// VeriBlock PoP Miner
// Copyright 2017-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.miners.pop.rules.actions;

import com.google.inject.Inject;
import nodecore.miners.pop.InternalEventBus;
import nodecore.miners.pop.contracts.MineResult;
import nodecore.miners.pop.contracts.PoPMiner;
import nodecore.miners.pop.contracts.ResultMessage;
import nodecore.miners.pop.events.ErrorMessageEvent;
import org.apache.commons.lang3.StringUtils;

public class MineAction implements RuleAction<Integer> {
    private final PoPMiner miner;

    @Inject
    public MineAction(PoPMiner miner) {
        this.miner = miner;
    }

    @Override
    public void execute(Integer blockHeight) {
        MineResult result = miner.mine(blockHeight);
        if (result.didFail()) {
            StringBuilder errorMessage = new StringBuilder();
            for (ResultMessage message : result.getMessages()) {
                errorMessage.append(System.lineSeparator())
                        .append(message.getMessage())
                        .append(": ")
                        .append(StringUtils.join(message.getDetails(), "; "));
            }
            InternalEventBus.getInstance().post(new ErrorMessageEvent(String.format("Mine Action Failed: %s", errorMessage.toString())));
        }
    }
}

File: nodecore-miners-pop/src/main/java/nodecore/miners/pop/rules/actions/MineAction.kt
// VeriBlock PoP Miner
// Copyright 2017-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.miners.pop.rules.actions

import nodecore.miners.pop.MinerService
import org.apache.commons.lang3.StringUtils
import org.veriblock.core.utilities.createLogger

private val logger = createLogger {}

class MineAction(
    private val minerService: MinerService
) : RuleAction<Int> {
    override fun execute(value: Int?) {
        // The given value is the block index to mine (or null)
        val result = minerService.mine(value)
        if (result.didFail()) {
            val errorMessage = StringBuilder()
            for (message in result.messages) {
                errorMessage.append(System.lineSeparator())
                    .append(message.message)
                    .append(": ")
                    .append(StringUtils.join(message.details, "; "))
            }

            logger.error { "Mine Action Failed: $errorMessage" }
        }
    }
}
